CI: Specify libc and libstdcxx versions in conda environments

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121568
* #121733
* #121268
* __->__ #121556

Without this we get mismatches between the GLIBC and GLIBCXX ABI used
by conda packages vs pytorch.

peterbell10 committed:
CI: Specify sysroot_linux-64 version in conda environments

Without this we get mismatches between the GLIBC and GLIBCXX ABI used
by conda packages vs pytorch.

[ghstack-poisoned]

isuruf commented on .ci/docker/common/install_conda.sh:
```diff
@@ -58,7 +58,10 @@ fi
 
   # Install correct Python version
-  as_jenkins conda create -n py_$ANACONDA_PYTHON_VERSION -y python="$ANACONDA_PYTHON_VERSION"
+  # Also ensure sysroot is using a modern GLIBC and GLIBCXX to match system compilers
```
This just updates GLIBC. Nothing to do with GLIBCXX
```suggestion
  # Also ensure sysroot is using a modern GLIBC to match system compilers
```

peterbell10 committed:
Update on "CI: Specify sysroot_linux-64 version in conda environments"


Without this we get mismatches between the GLIBC and GLIBCXX ABI used
by conda packages vs pytorch.

[ghstack-poisoned]

peterbell10 committed:
Update on "CI: Specify sysroot_linux-64 version in conda environments"


Without this we get mismatches between the GLIBC and GLIBCXX ABI used
by conda packages vs pytorch.

[ghstack-poisoned]

isuruf commented on .ci/docker/common/install_conda.sh:
```diff
@@ -58,7 +58,14 @@ fi
+  # Also ensure sysroot is using a modern GLIBC to match system compilers
+  as_jenkins conda create -n py_$ANACONDA_PYTHON_VERSION -y\
+             python="$ANACONDA_PYTHON_VERSION" \
+             sysroot_linux-64=2.17
```
Since this script supports aarch64, this needs to change depend on `uname -m`

peterbell10 committed:
Update

[ghstack-poisoned]

malfet commented on .ci/docker/common/install_conda.sh:
```diff
@@ -57,8 +57,21 @@ fi
+             python="$ANACONDA_PYTHON_VERSION" \
+             ${SYSROOT_DEP}
+
+  # libstdcxx from conda default channels are too old, we need GLIBCXX_3.4.30
+  # which is provided in libstdcxx 12 and up.
+  conda_install libstdcxx-ng=12.3.0 -c conda-forge
```
Problem with this statement, is that it adds conda-forge channel, which allows the rest of the packages to be installed from conda-forge going forward. But may be this is fine

malfet commented:
@pytorchbot merge

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

atalman commented:
@pytorchbot cherry-pick --onto release/2.3 -c regression --fixes "glibc errors"
